lib/bolt/transport/orch.rb

```diff
@@ -0,0 +1,118 @@
+require 'base64'
+require 'json'
+require 'orchestrator_client'
+
+module Bolt
+  module Transport
+    class Orch
+      CONF_FILE = File.expand_path('~/.puppetlabs/client-tools/orchestrator.conf')
+      BOLT_MOCK_FILE = 'bolt/tasks/init'.freeze
+
+      attr_reader :client, :logger
+
+      def initialize(config)
+        client_keys = %i[service-url token-file cacert]
+        client_opts = config.select { |k, _v| client_keys.include?(k) }
+        @client = OrchestratorClient.new(client_opts, true)
+        @logger = Logging.logger[self]
+      end
+
+      def upload(target, source, destination, _options = {})
+        content = File.open(source, &:read)
+        content = Base64.encode64(content)
+        mode = File.stat(source).mode
+        params = {
+          action: 'upload',
+          path: destination,
+          content: content,
+          mode: mode
+        }
+        result = run_task(target, BOLT_MOCK_FILE, 'stdin', params)
+        result = Bolt::Result.for_upload(target, source, destination) unless result.error_hash
+        result
+      end
+
+      def run_command(target, command, _options = {})
+        result = run_task(target,
+                          BOLT_MOCK_FILE,
+                          'stdin',
+                          action: 'command',
+                          command: command,
+                          options: {})
+        unwrap_bolt_result(target, result)
+      end
+
+      def run_script(target, script, arguments, _options = {})
+        content = File.open(script, &:read)
+        content = Base64.encode64(content)
+        params = {
+          action: 'script',
+          content: content,
+          arguments: arguments
+        }
+        unwrap_bolt_result(target, run_task(target, BOLT_MOCK_FILE, 'stdin', params))
+      end
+
+      def run_task(target, task, _inputmethod, arguments, _options = {})
+        body = { task: task_name_from_path(task),
+                 environment: target.options[:orch_task_environment],
+                 noop: arguments['_noop'],
+                 params: arguments.reject { |k, _| k == '_noop' },
+                 scope: {
+                   nodes: [target.host]
+                 } }
+        # Should we handle errors here or let them propagate?
```

Comment:
Now is the time to answer this. I think we can leave it as is and remove this comment